File: dotnet/Samples/EndUserSample/EndUserSample.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Text.Json;
using Jose;
using MinaOmbud;
using MinaOmbud.Models;

var defaults = Defaults.Instance;
var authUrl = defaults.MINA_OMBUD_API_TOKEN_URL;
var apiUrl = defaults.MINA_OMBUD_API_URL;
var serializeOptions = JsonCanonicalizer.SerializerOptions;
var acceptableKeyTypes = new HashSet<string> { Jwk.KeyTypes.RSA };
var acceptableCryptos = new HashSet<string> { "RS256", "RS384", "RS512" };
var unauthenticatedClient = new HttpClient();
var authenticatedClient = new HttpClient();
authenticatedClient.Timeout = Timeout.InfiniteTimeSpan;

///////////////////////////////////////////////////////////////////////////////
// 1. User claims
// These are the values identifying the user.

// Issue and expiry times (2 minutes)
var iat = DateTimeOffset.Now.ToUnixTimeSeconds();
var exp = iat + 60 * 2;

const string ssn = "198602262381"; // Social security number
var userClaims = new Dictionary<string, object>
{
    { "https://claims.oidc.se/1.0/personalNumber", ssn },
    { "name", "Beri Ylles" },
    { "given_name", "Beri" },
    { "family_name", "Ylles" },
    { "iat", iat },
    { "exp", exp },
    { "iss", "http://localhost" },
    { "aud", "mina-ombud" },
};

///////////////////////////////////////////////////////////////////////////////
// 2. Sign user claims
///////////////////////////////////////////////////////////////////////////////

// a) Load signing key
var p12 = new X509Certificate2(Path.Join(defaults.MINA_OMBUD_SAMPLE_DATA, "keys/signing.p12"), "",
    X509KeyStorageFlags.Exportable | X509KeyStorageFlags.MachineKeySet);

// b) Construct a signed token
var userToken = JWT.Encode(userClaims, p12.GetRSAPrivateKey(), JwsAlgorithm.RS256);

///////////////////////////////////////////////////////////////////////////////
// 3. Request API access token.
// The access token should be requested and reused for subsequent requests
// until it expires at which point a new token must be requested.

var tokenRequest = new FormUrlEncodedContent(new Dictionary<string, string>
    {
        { "grant_type", "client_credentials" },
        { "client_id", defaults.MINA_OMBUD_API_CLIENT_ID },
        { "client_secret", defaults.MINA_OMBUD_API_CLIENT_SECRET },
        { "scope", "user:self" }
    }
);

var tokenResponse = await Post<Dictionary<string, object>>(authUrl, tokenRequest);
var accessToken = tokenResponse["access_token"].ToString();
authenticatedClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessToken}");
authenticatedClient.DefaultRequestHeaders.Add("x-id-token", userToken);
authenticatedClient.DefaultRequestHeaders.Add("x-service-name", "EndUserSample.cs");
authenticatedClient.DefaultRequestHeaders.Add("x-request-id", Guid.NewGuid().ToString());

///////////////////////////////////////////////////////////////////////////////
// 4. Invoke API
var request = new HamtaBehorigheterRequest()
{
    Tredjeman = "2120000829", // Where the permission is exercised
    Fullmaktshavare = new Identitetsbeteckning(ssn), // Holder of the permission
    Fullmaktsgivarroll = new List<string> { "ORGANISATION" }, // Filter on issuer type
    Page = new PageParameters(0, 100), // Pagination
    // Request permissions for specific issuer
    // Fullmaktsgivare = new Identitetsbeteckning("556...", "orgnr"),
    // Filter on specific permissions
    // Behorigheter = new List<string> { "5611f2d8-c74e-46e4-aab1-b2f0bd4ce318" }
};

var response = await ApiPost<HamtaBehorigheterResponse>($"{apiUrl}/sok/behorigheter", request);
Console.WriteLine(PrettyPrint(response));

///////////////////////////////////////////////////////////////////////////////
// 5. Verify response signature and timestamp
// When the permissions are passed on to other services/systems
// instead of being used right away it is important to verify
// the digital signature of the permissions in the receiving
// service.
//
// This ensures the permissions have not been tampered with.
//
// This verification should take place in the receiving service.
foreach (var kontext in response.Kontext)
{
    var givare = kontext.Fullmaktsgivare;
    var havare = kontext.Fullmaktshavare[0];
    Console.WriteLine($"=== fullmaktsgivare={givare.Namn}, fullmaktshavare={havare.Fornamn} {havare.Namn} ===");
    // a) Fetch key set
    //    In a real implementation the keys would be cached
    //    and only fetched when a new key is used.
    var keys = await FetchJwkSet(kontext.Tredjeman);

    // b) Produce the canonical JSON representation of the payload
    //    without the embedded signature.
    var payload = JsonCanonicalizer.Serialize(kontext, detachSig: true);
    Console.WriteLine(payload);

    // c) Encode the JSON string as UTF-8 and convert to Base64 URL without padding
    var b64 = Base64Url.Encode(Encoding.UTF8.GetBytes(payload));

    // d) Construct the JWS
    var jws = kontext.Sig.Protected + "." + b64 + "." + kontext.Sig.Signature;

    // e) Verify header and find the public key to use for verification.
    var headers = JWT.Headers(jws);
    var keyId = (string)headers["kid"];
    var alg = (string)headers["alg"];
    if (!acceptableCryptos.Contains(alg))
    {
        throw new CryptographicException("Unsupported signing algorithm");
    }

    var pubKey = (
        from key in keys
        where key.KeyId == keyId
              && acceptableKeyTypes.Contains(key.Kty)
              && (key.Alg == null || acceptableCryptos.Contains(key.Alg))
        select key
    ).First();

    // f) Verify signature
    JWT.Decode(jws, pubKey);

    // g) Verify timestamp.
    // The tolerance is very application dependent.
    // Here we accept information no older than 2 minutes.
    var timestamp = DateTimeOffset.Parse(kontext.Tidpunkt, CultureInfo.InvariantCulture);
    var delta = DateTimeOffset.Now.Subtract(timestamp);
    if (delta.TotalMinutes > 2)
    {
        Console.WriteLine($"Expired: {timestamp} : {delta}");
    }
}

///////////////////////////////////////////////////////////////////////////////
// Utilities
///////////////////////////////////////////////////////////////////////////////
static string PrettyPrint(object value)
{
    return JsonSerializer.Serialize(value, new JsonSerializerOptions(JsonCanonicalizer.SerializerOptions)
    {
        WriteIndented = true
    });
}

async Task<JwkSet> FetchJwkSet(string tredjeman)
{
    var json = await unauthenticatedClient.GetStringAsync($"{apiUrl}/tredjeman/{tredjeman}/jwks");
    return JwkSet.FromJson(json, JWT.DefaultSettings.JsonMapper);
}

async Task<T> Post<T>(string uri, HttpContent value)
{
    var r = await unauthenticatedClient.PostAsync(uri, value);
    r.EnsureSuccessStatusCode();
    var body = await r.Content.ReadFromJsonAsync<T>(serializeOptions);
    if (body == null)
    {
        throw new InvalidDataException($"No response body returned from {uri}");
    }

    return body;
}

async Task<T> ApiPost<T>(string uri, object value)
{
    var r = await authenticatedClient.PostAsJsonAsync(uri, value, serializeOptions);
    r.EnsureSuccessStatusCode();
    var body = await r.Content.ReadFromJsonAsync<T>(serializeOptions);
    if (body == null)
    {
        throw new InvalidDataException($"No response body returned from {uri}");
    }

    return body;
}

File: python/src/minaombud/samples/enduser_sample.py
#
# Sample demonstrating using the API acting as an end user
# that is a party to the Power of attorney.
#
# This is the standard case.
# A user wants to act using the permissions granted by a Power of attorney.
#
import base64
import functools
import json
import sys
import time
import uuid
from datetime import datetime

import requests

from minaombud import defaults
from minaombud.crypto import jose
from minaombud.crypto.jose import JOSEHeader
from minaombud.crypto.jwkset import JwkSet, RemoteJwkSet
from minaombud.model import JwsSig
from minaombud.util import parse_iso_datetime


def sample():
    ### 1. User claims
    iat = int(time.time())  # Time of issue
    exp = iat + 60 * 2      # Expiry time 2 minutes
    ssn = "198602262381"    # Social security number
    user_claims = {
        "https://claims.oidc.se/1.0/personalNumber": ssn,
        # "https://claims.oidc.se/1.0/coordinationNumber": ssn,
        "name": "Beri Ylles",
        "given_name": "Beri",
        "family_name": "Ylles",
        "iat": iat,
        "exp": exp,
        "iss": "http://localhost",
        "aud": "mina-ombud",
        "sub": "9ebe70e4-ca61-11ed-97ed-00155d52ccdb"
    }

    ### 2. Sign claims
    # a) Load signing key
    key_set = JwkSet.load(defaults.MINA_OMBUD_SAMPLE_KEYS)
    try:
        key = next(k for k in key_set.private_keys)
        assert key.kty == "RSA"
    except StopIteration:
        raise KeyError("No private key found")

    # b) Sign claims to get a JWS using compact serialization
    header = JOSEHeader(alg="RS256", kid=key.kid)
    user_token = jose.sign(user_claims, key, header)

    ### 3. Request API access token
    # The access token should be requested and reused for subsequent requests
    # until it expires at which point a new token must be requested.
    client_id = defaults.MINA_OMBUD_API_CLIENT_ID
    client_secret = defaults.MINA_OMBUD_API_CLIENT_SECRET
    token_url = defaults.MINA_OMBUD_API_TOKEN_URL
    token_request = {
        "grant_type": "client_credentials",
        "client_id": client_id,
        "client_secret": client_secret,
        "scope": "user:self",
    }
    request_time = time.time()
    token_response = requests.post(token_url, token_request).json()
    assert token_response["token_type"] == "Bearer"
    access_token = token_response["access_token"]
    token_expires_at = request_time + token_response["expires_in"]

    ### 4. Invoke API
    api_url = defaults.MINA_OMBUD_API_URL
    behorigheter_request = {
        "tredjeman": defaults.MINA_OMBUD_TREDJE_MAN,  # Where the permission is used
        "fullmaktshavare": {"id": ssn, "typ": "pnr"},  # Holder of the permission
        "fullmaktsgivarroll": ["ORGANISATION"],  # Filter on issuer type
        # "fullmaktsgivare": { "id": "556...", "typ": "orgnr" },# Filter on issuer of permissions
        # "behorigheter": [                                     # Filter on specific permissions
        #     "ac94b31e-a17f-11ed-b19d-00155d41fac2"
        # ],
        "page": {"page": 0, "size": 100},  # Pagination
    }

    headers = {
        "authorization": f"Bearer {access_token}",
        "x-id-token": user_token,
        "x-service-name": "enduser_sample",
        "x-request-id": str(uuid.uuid4()),
    }
    behorigheter_response = requests.post(
        f"{api_url}/sok/behorigheter", json=behorigheter_request, headers=headers
    )
    content_type = behorigheter_response.headers.get("content-type")
    if content_type != "application/json":
        for kontext, v in behorigheter_response.headers.items():
            print(f"{kontext}: {v}", file=sys.stderr)
        print(file=sys.stderr)
        print(behorigheter_response.text, file=sys.stderr)
        return

    print(json.dumps(behorigheter_response.json(), indent=2))

    ### 5. Verify response signature and timestamp
    # When the permissions are passed on to other services/systems
    # instead of being used right away it is important to verify
    # the digital signature of the permissions in the receiving
    # service.
    #
    # This ensures the permissions have not been tampered with.
    #
    # This verification should take place in the receiving service.
    perr = functools.partial(print, file=sys.stderr)
    for kontext in behorigheter_response.json()["kontext"]:
        # a) Fetch key set
        #    In a real implementation the keys would be cached
        #    and only fetched when a new key is used.
        tredjeman: str = kontext["tredjeman"]
        assert tredjeman.isdigit()
        key_set = RemoteJwkSet(f"{api_url}/tredjeman/{tredjeman}/jwks")

        # b) Detach the embedded signature from the signed object.
        sig = kontext["_sig"]
        del kontext["_sig"]

        # c) Produce the canonical JSON representation of the payload
        canonical = json.dumps(
            kontext,
            ensure_ascii=False,
            allow_nan=False,
            sort_keys=True,
            separators=(",", ":"),
        )

        # d) Encode UTF-8 representation of the JSON payload as Base 64-url
        payload = base64.urlsafe_b64encode(canonical.encode("utf-8")).rstrip(b"=")

        # e) Verify the detached signature against the canonical payload:
        #    sig["signature"] == signature(sig["protected"] + '.' + payload)
        valid = jose.verify_jws(payload, JwsSig.from_dict(sig), key_set)
        if not valid:
            perr("Signature verification failed")
            perr(f"- Header    : {sig['protected']}")
            perr(f"- Payload   : {payload.decode('ascii')}")
            perr(f"- Signature : {sig['signature']}")
            perr(json.dumps(kontext, ensure_ascii=False, indent=2, sort_keys=True))

        # f) Check timestamp of permissions.
        #    The tolerance is very application dependent.
        #    Here we accept up to 2 minutes old information.
        iso_timestamp = kontext["tidpunkt"]
        timestamp = parse_iso_datetime(iso_timestamp)
        now = datetime.now(timestamp.tzinfo)
        delta = now - timestamp
        if delta.total_seconds() > 2 * 60:
            perr(f"Expired: {iso_timestamp} : {delta}")
            perr(json.dumps(kontext, ensure_ascii=False, indent=2, sort_keys=True))


if __name__ == "__main__":
    sample()
